File: plugins/autoresearch/hooks/result-regression-check.js
#!/usr/bin/env node
/**
 * PostToolUse Hook: Result Regression Check
 *
 * After editing train.py (or similar training scripts), reminds the agent
 * to run an experiment before committing the change.
 *
 * Non-blocking (exit code 0).
 */

const fs = require('fs');
const path = require('path');
const MAX_STDIN = 1024 * 1024;

const TRAINING_FILES = ['train.py', 'training.py', 'train_model.py'];

let data = '';
process.stdin.setEncoding('utf8');

process.stdin.on('data', chunk => {
  if (data.length < MAX_STDIN) data += chunk.substring(0, MAX_STDIN - data.length);
});

process.stdin.on('end', () => {
  try {
    const disabled = (process.env.DISABLED_HOOKS || '').split(',');
    if (disabled.includes('result-regression-check')) {
      process.stdout.write(data);
      return;
    }

    const input = JSON.parse(data);
    const filePath = input.tool_input?.file_path || '';
    const basename = path.basename(filePath);

    const isTrainingFile = TRAINING_FILES.some(f => basename === f) ||
                           /train.*\.py$/.test(basename);

    if (!isTrainingFile) {
      process.stdout.write(data);
      return;
    }

    const logPath = path.join(process.cwd(), '.autoresearch', 'experiments', 'experiment-log.jsonl');
    let lastRunTime = 0;
    try {
      const stat = fs.statSync(logPath);
      lastRunTime = stat.mtimeMs;
    } catch {}

    const fileModTime = Date.now();
    const timeSinceLastRun = fileModTime - lastRunTime;
    const fiveMinutes = 5 * 60 * 1000;

    if (timeSinceLastRun > fiveMinutes || lastRunTime === 0) {
      console.error(`[Hook] REMINDER: You edited ${basename}. Run an experiment before committing.`);
      console.error(`[Hook] Use: /experiment run — or: uv run train.py`);
      if (lastRunTime === 0) {
        console.error('[Hook] No experiment log found. Run /experiment baseline to establish a baseline first.');
      }
    }

    process.stdout.write(data);
  } catch (err) {
    console.error(`[Hook] result-regression-check error: ${err.message}`);
    process.stdout.write(data);
  }
});

File: plugins/autoresearch/hooks/result-regression-check.py
#!/usr/bin/env python3
"""
PostToolUse Hook: Result Regression Check

After editing train.py (or similar training scripts), reminds the agent
to run an experiment before committing the change.

Non-blocking (exit code 0).
"""

import json
import os
import re
import sys
import time

MAX_STDIN = 1024 * 1024

TRAINING_FILES = ["train.py", "training.py", "train_model.py"]


def main():
    data = sys.stdin.read(MAX_STDIN)

    try:
        disabled = os.environ.get("DISABLED_HOOKS", "").split(",")
        if "result-regression-check" in disabled:
            sys.stdout.write(data)
            return

        inp = json.loads(data)
        file_path = inp.get("tool_input", {}).get("file_path", "")
        basename = os.path.basename(file_path)

        is_training_file = (
            basename in TRAINING_FILES
            or bool(re.search(r"train.*\.py$", basename))
        )

        if not is_training_file:
            sys.stdout.write(data)
            return

        log_path = os.path.join(os.getcwd(), ".autoresearch", "experiments", "experiment-log.jsonl")
        last_run_time = 0
        try:
            last_run_time = os.stat(log_path).st_mtime * 1000
        except (FileNotFoundError, OSError):
            pass

        file_mod_time = time.time() * 1000
        time_since_last_run = file_mod_time - last_run_time
        five_minutes = 5 * 60 * 1000

        if time_since_last_run > five_minutes or last_run_time == 0:
            print(f"[Hook] REMINDER: You edited {basename}. Run an experiment before committing.", file=sys.stderr)
            print("[Hook] Use: /experiment run — or: uv run train.py", file=sys.stderr)
            if last_run_time == 0:
                print("[Hook] No experiment log found. Run /experiment baseline to establish a baseline first.", file=sys.stderr)

        sys.stdout.write(data)
    except Exception as err:
        print(f"[Hook] result-regression-check error: {err}", file=sys.stderr)
        sys.stdout.write(data)


if __name__ == "__main__":
    main()
